File: frontend/lib/models/booking.dart
class Booking {
  final String id;
  final String customerId;
  final String serviceId;
  final String providerId;
  final CustomerDetails customerDetails;
  final DateTime reservationDate;
  final String status;
  final double totalAmount;
  final int hoursBooked;
  final String bookingType; // 'immediate' or 'reservation'
  final String? cancellationReason;
  final String? cancelledBy;
  final bool canCancel;
  final DateTime createdAt;
  final DateTime updatedAt;
  
  // Completion confirmation tracking
  final String? completionInitiatedBy;
  final DateTime? completionInitiatedAt;
  final String? completionConfirmedBy;
  final DateTime? completionConfirmedAt;
  
  // Rating tracking
  final bool customerRated;
  final bool providerRated;
  
  // Populated fields from backend
  final String? customerName;
  final String? providerName;
  final String? serviceName;

  Booking({
    required this.id,
    required this.customerId,
    required this.serviceId,
    required this.providerId,
    required this.customerDetails,
    required this.reservationDate,
    required this.status,
    required this.totalAmount,
    required this.hoursBooked,
    this.bookingType = 'reservation',
    this.cancellationReason,
    this.cancelledBy,
    required this.canCancel,
    required this.createdAt,
    required this.updatedAt,
    this.completionInitiatedBy,
    this.completionInitiatedAt,
    this.completionConfirmedBy,
    this.completionConfirmedAt,
    this.customerRated = false,
    this.providerRated = false,
    this.customerName,
    this.providerName,
    this.serviceName,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    // Helper function to extract ID from string or object
    String extractId(dynamic value) {
      if (value == null) return '';
      if (value is String) return value;
      if (value is Map) return value['_id']?.toString() ?? '';
      return value.toString();
    }
    
    // Helper function to extract name from populated user object
    String? extractName(dynamic value) {
      if (value == null) return null;
      if (value is Map) {
        final profile = value['profile'];
        if (profile is Map) {
          return profile['name']?.toString();
        }
      }
      return null;
    }
    
    // Helper function to extract service name
    String? extractServiceName(dynamic value) {
      if (value == null) return null;
      if (value is Map) {
        return value['serviceName']?.toString();
      }
      return null;
    }

    return Booking(
      id: json['_id']?.toString() ?? '',
      customerId: extractId(json['customerId']),
      serviceId: extractId(json['serviceId']),
      providerId: extractId(json['providerId']),
      customerDetails: CustomerDetails.fromJson(
        json['customerDetails'] is Map 
            ? Map<String, dynamic>.from(json['customerDetails'] as Map)
            : {}
      ),
      reservationDate: DateTime.parse(
        json['reservationDate'] ?? DateTime.now().toIso8601String()
      ),
      status: json['status']?.toString() ?? 'pending',
      totalAmount: (json['totalAmount'] ?? 0).toDouble(),
      hoursBooked: (json['hoursBooked'] ?? 1) is int 
          ? json['hoursBooked'] 
          : int.tryParse(json['hoursBooked'].toString()) ?? 1,
      bookingType: json['bookingType']?.toString() ?? 'reservation',
      cancellationReason: json['cancellationReason']?.toString(),
      cancelledBy: json['cancelledBy']?.toString(),
      canCancel: json['canCancel'] == true,
      createdAt: DateTime.parse(
        json['createdAt'] ?? DateTime.now().toIso8601String()
      ),
      updatedAt: DateTime.parse(
        json['updatedAt'] ?? DateTime.now().toIso8601String()
      ),
      completionInitiatedBy: json['completionInitiatedBy']?.toString(),
      completionInitiatedAt: json['completionInitiatedAt'] != null
          ? DateTime.parse(json['completionInitiatedAt'])
          : null,
      completionConfirmedBy: json['completionConfirmedBy']?.toString(),
      completionConfirmedAt: json['completionConfirmedAt'] != null
          ? DateTime.parse(json['completionConfirmedAt'])
          : null,
      customerRated: json['customerRated'] == true,
      providerRated: json['providerRated'] == true,
      customerName: extractName(json['customerId']),
      providerName: extractName(json['providerId']),
      serviceName: extractServiceName(json['serviceId']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'customerId': customerId,
      'serviceId': serviceId,
      'providerId': providerId,
      'customerDetails': customerDetails.toJson(),
      'reservationDate': reservationDate.toIso8601String(),
      'status': status,
      'totalAmount': totalAmount,
      'hoursBooked': hoursBooked,
      'cancellationReason': cancellationReason,
      'cancelledBy': cancelledBy,
      'canCancel': canCancel,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class CustomerDetails {
  final String name;
  final String phoneNumber;
  final String exactAddress;

  CustomerDetails({
    required this.name,
    required this.phoneNumber,
    required this.exactAddress,
  });

  factory CustomerDetails.fromJson(Map<String, dynamic> json) {
    return CustomerDetails(
      name: json['name']?.toString() ?? '',
      phoneNumber: json['phoneNumber']?.toString() ?? '',
      exactAddress: json['exactAddress']?.toString() ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'phoneNumber': phoneNumber,
      'exactAddress': exactAddress,
    };
  }
}

File: backend/models/Booking.js
const mongoose = require('mongoose');

const bookingSchema = new mongoose.Schema({
  customerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  serviceId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Service',
    required: true
  },
  providerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  customerDetails: {
    name: {
      type: String,
      required: true
    },
    phoneNumber: {
      type: String,
      required: true
    },
    exactAddress: {
      type: String,
      required: true
    }
  },
  bookingType: {
    type: String,
    enum: ['immediate', 'reservation'],
    default: 'reservation',
    required: true
  },
  reservationDate: {
    type: Date,
    required: false,
    default: null
  },
  status: {
    type: String,
    enum: ['pending', 'confirmed', 'in_progress', 'completed', 'cancelled'],
    default: 'pending'
  },
  totalAmount: {
    type: Number,
    required: true
  },
  hoursBooked: {
    type: Number,
    required: true,
    default: 1
  },
  cancellationReason: String,
  cancelledBy: {
    type: String,
    enum: ['customer', 'provider']
  },
  canCancel: {
    type: Boolean,
    default: true
  },
  conversationId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Chat'
  },
  notificationSent: {
    type: Boolean,
    default: false
  },
  acceptedAt: {
    type: Date
  },
  // Completion confirmation tracking
  completionInitiatedBy: {
    type: String,
    enum: ['customer', 'provider']
  },
  completionInitiatedAt: {
    type: Date
  },
  completionConfirmedBy: {
    type: String,
    enum: ['customer', 'provider']
  },
  completionConfirmedAt: {
    type: Date
  },
  // Rating tracking
  customerRated: {
    type: Boolean,
    default: false
  },
  providerRated: {
    type: Boolean,
    default: false
  }
}, {
  timestamps: true
});

// Add indexes for performance optimization
bookingSchema.index({ conversationId: 1 });
bookingSchema.index({ customerId: 1, providerId: 1 });
bookingSchema.index({ status: 1 });
bookingSchema.index({ reservationDate: 1 });

// Middleware to check if booking can be cancelled (3 hours before reservation)
bookingSchema.pre('save', function (next) {
  const now = new Date();
  const reservationTime = new Date(this.reservationDate);
  const timeDiff = reservationTime.getTime() - now.getTime();
  const hoursDiff = timeDiff / (1000 * 3600);

  this.canCancel = hoursDiff > 3;
  next();
});

module.exports = mongoose.model('Booking', bookingSchema);